gql/comparator.go

```diff
@@ -0,0 +1,207 @@
+	}
+
+	values := make([]string, value.Len())
+	for i := 0; i < value.Len(); i++ {
+		rawItem := value.Index(i).Interface()
+		item, ok := rawItem.(string)
+		if !ok {
+			return nil, errors.New("filter argument 'Values' should be a list of strings")
+		}
+		values[i] = item
+	}
+
+	return inComparator{values: values}, nil
+}
+
+// NewIntegerComparator returns a NewListOperator for integer operations supported by integerComparator,
+// specifically <, <=, > and >=.
+// The returned NewListOperation expects an int with the key "Value" as part of the given argument.
+func NewIntegerComparator(op string) NewListOperation {
+	switch op {
+	case ">", ">=", "<", "<=":
+		break
+	default:
+		return func(arg map[string]interface{}) (Comparator, error) {
+			return nil, fmt.Errorf("unsupported integer comparison operation %q", op)
+		}
+	}
+	return func(arg map[string]interface{}) (Comparator, error) {
+		raw, ok := arg["Value"]
+		if !ok {
+			return nil, errors.New("filter argument is missing 'Value'")
+		}
+		value, ok := raw.(int)
+		if !ok {
+			return nil, errors.New("filter argument 'Value' must be an integer")
+		}
+
+		return integerComparator{operation: op, value: value}, nil
+	}
+}
+
+// NewNotInComparator wraps NewInComparator returning the opposite boolean.
+func NewNotInComparator(arg map[string]interface{}) (Comparator, error) {
+	in, err := NewInComparator(arg)
+	if err != nil {
+		return nil, err
+	}
+	return NotComparator{Child: in}, nil
+}
+
+// NewLimitLengthComparator returns a comparator which simply limits the length of the results to the integer
+// given with the key "Value" in the argument.
+func NewLimitLengthComparator(arg map[string]interface{}) (Comparator, error) {
+	raw, ok := arg["Value"]
+	if !ok {
+		return nil, errors.New("filter argument is missing 'Value'")
+	}
+	limit, ok := raw.(int)
+	if !ok {
+		return nil, errors.New("filter argument 'Value' must be an integer")
+	}
+
+	return &limitLength{limit: limit}, nil
+}
+
+type inComparator struct {
+	values []string
+}
+
+func (c inComparator) Match(raw interface{}) bool {
+	in, ok := raw.(string)
+	if !ok {
+		return false
+	}
+	for _, v := range c.values {
+		if in == v {
+			return true
+		}
+	}
+	return false
+}
+
+type intEqual struct {
+	value int
+}
+
+func (c intEqual) Match(raw interface{}) bool {
+	in, ok := raw.(int)
+	if !ok {
+		return false
+	}
+	return in == c.value
+}
+
+// integerComparator supports comparisons between ints for a these operations <, <=, > and >=.
+type integerComparator struct {
+	operation string
+	value     int
+}
+
+func (c integerComparator) Match(raw interface{}) bool {
+	in, ok := raw.(int)
+	if !ok {
+		return false
+	}
+	switch c.operation {
+	case ">":
+		return in > c.value
+	case ">=":
+		return in >= c.value
+	case "<":
+		return in < c.value
+	case "<=":
+		return in <= c.value
+	default:
+		return false
+	}
+}
+
+type limitLength struct {
+	limit int
+	count int
+	mtx   sync.Mutex
```

Comment:
Since `ResolveListField` does the matches sequentially, I wouldn't think you would need to have a mutex here.  If you *do* need a mutex here, how confident are you the `count` is accurate and not being shared between unrelated filter lists?

Reply:
The parallel handling of queries in the GraphQL library we are currently using is a bit of a mess, however in general when it hits a list the resolving of each list item is done in its own go routine. That allows for parallel retrieval in the case of the resolving of a field requiring a DB call. Basically that means I do think I need a mutex here.

Regarding how confident I am it is not being shared between unrelated filter lists I'm confident. The comparator is made as part of the filter which is created in ResolveListField which runs fresh for each query being processed and multiple times depending on which field it is associated with.

Reply:
Given the place where you have inserted this resolve function, shouldn't all the items in that list have already been finished by the time you're doing the match processing? So the limit itself wouldn't need to be parallel because all the parallel bits happen lower and are done?

Reply:
hmm, maybe that is right the parallel bits have already been done. Though that sounds right that code is complicated and this is usable by other uses cases since NewLimitLengthComparator is public. That makes me inclined to keep the mutex for the safety it brings.

Reply:
Sounds reasonable to me.

I worry, however, that anything that does take advantage of the mutex will also have to deal with a non-deterministic return value.